lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs

```diff
@@ -36,31 +34,30 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
     /// </summary>
     /// <typeparam name="TMapInput"></typeparam>
     /// <typeparam name="TMapOutput"></typeparam>
-    internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutput>
+    /// <typeparam name="TDataHandler"></typeparam>
+    internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutput, TDataHandler>
     {
-        private static readonly Logger Logger = Logger.GetLogger(typeof(ServiceAndContextConfigurationProvider<TMapInput, TMapOutput>));
+        private static readonly Logger Logger = Logger.GetLogger(typeof(ServiceAndContextConfigurationProvider<TMapInput, TMapOutput, TDataHandler>));
 
-        private readonly Dictionary<string, ContextAndServiceConfiguration> _configurationProvider;
-        private readonly ISet<string> _failedEvaluators;
-        private readonly ISet<string> _submittedEvaluators; 
+        private readonly Dictionary<string, string> _configurationProvider;
+        private readonly ISet<string> _submittedEvaluators;
+        private readonly ISet<string> _contextLoadedEvaluators; 
         private readonly object _lock;
-        private readonly int _numNodes;
-        private int _assignedPartitionDescriptors;
-        private readonly IGroupCommDriver _groupCommDriver;
-        private readonly ConfigurationManager _configurationManager;
-        private readonly Stack<IPartitionDescriptor> _partitionDescriptors;
-
-        internal ServiceAndContextConfigurationProvider(int numNodes, IGroupCommDriver groupCommDriver,
-            ConfigurationManager configurationManager, Stack<IPartitionDescriptor> partitionDescriptors)
+        private readonly Stack<string> _partitionDescriptorIds;
+        private readonly IPartitionedInputDataSet _dataset;
+        private string _masterEvaluatorId;
```

Comment:
I would prefer `Optional` in cases where such variables can be `null` and/or reassigned.